### ISSUE TITLE
Promise.all and Promise.allSettled infer incorrect types from Array subclasses

### ISSUE THREAD

**Issue by asnaeb** [microsoft/TypeScript#51993]
# lib Update Request
## Configuration Check
My compilation *target* is `ES2020` and my *lib* is `the default`.

## Missing / Incorrect Definition

`Promise.all`
`Promise.allSettled`

## Sample Code
```ts
declare class ArrayChild<T> extends Array<T> {}

declare const array: Array<Promise<number>>
declare const child: ArrayChild<Promise<number>>

Promise.all(array).then(a => a /* number[] */) 
Promise.allSettled(array).then(a => a /* PromiseSettledResult<number>[] */)

Promise.all(child).then(a => a /* {[K in keyof ArrayChild]: number} */) 
Promise.allSettled(child).then(a => /* {[K in keyof ArrayChild]: PromiseSettledResult<number>} */) 
```
[Playground link](https://www.typescriptlang.org/play?target=7#code/CYUwxgNghgTiAEkoGdnwIIxlAngYQAsBLCYAHgBUA+eEADwBcQA7YNTbHSmgbwF8AUANBI4iAPbNkDeLE4AuDFlxkACjHEBbIshBlmAV00AjEDCpVh4aGLCTpiYqUUdchEuXVadewybMWVqIIdlIyREzYxhAgijwA2gDKOCbiEAB0EWZQDOIwALqKABQAlPAAvDQAkpE5efpGpuYANPDxzIp+TYXwXtq6Df7mggJ9PulQEBBFcrgl6QwELDMVNFAlAgD0m-C7e-sHh4cAegD8oxr9IBNTiSAMDDHAM8o484vLUKuyG9tH-wDAWchH8atloggIOJxABrNAAc2hwHgF28uhu0yyURi7yWzBWlR+Wx2gNJ-2BY3RkwgdweTyKWKgENxn2+62JZM5ZOBIJ2rhwjg8iGgqHgwHEIGQzAA5DJ4fd4B94DAiPCCDIGDgAA5EZjw1FXDFFMBOYAs-FfQnsv5c7nnSnXam0x4gZ4mjzmglrX4k21++BnIA)
## Documentation Link
[Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)
[Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled)

## Possible fix
The inferred value should be of type `number[]` exactly like what happens with normal `Array` class. In the linked playground, It can be seen that `Promise.all` and `Promise.allSettled` infer the correct type with an `Iterable` type of object. This undesired behavior just happens with `Array` subclasses. I have quickly tried changing the signature from the original:
```ts
allSettled<T extends readonly unknown[] | []>(values: T): Promise<{ -readonly [P in keyof T]: PromiseSettledResult<Awaited<T[P]>> }>;
all<T extends readonly unknown[] | []>(values: T): Promise<{ -readonly [P in keyof T]: Awaited<T[P]> }>;
```
to
```ts
allSettled<T extends Promise<unknown>[]>(values: T): Promise<T extends Promise<infer S>[] ? PromiseSettledResult<S>[] : never>
all<T extends Promise<unknown>[]>(values: T): Promise<T extends Promise<infer S>[] ? S[] : never>
```
and it works as expected: [Playground link](https://www.typescriptlang.org/play?target=7#code/CYUwxgNghgTiAEkoGdnwIIxlAngYQAsBLCYAHgBUA+eEADwBcQA7YNTbHSmgbwF8AUANBI4iAPbNkDRMVIAuDFlyES5AAoxxAWyLIQZZgFdtAIxAwqVISOhiAZkeZgGRSfCgQIlWoxZt4TR09AycAa2ZxAHdmKgBtAF0qAAoAN08jEGRFCgBKRSDdfR96JlY0QpCyImZ7C3gAZXiE+AB+RsT4RWYQVItrW1gER2dXd08IBpAGBggQcgpfMoDK4vDImOaU9IhM7Pg8gq0ig0XS-wrjqpq6mEbmtsCr-SmZueAAJSyjCAYyJs63V6-SEE2SYDkwFyADoGAQWMkoPAALw0KC5AQAekx8FxePxBPxAD1WgIJq9ZvNwZCYXCEWAUTQwBjsYS2eyCSSBBygA)

**Comment by IllusionMH** [comment]
This would break tuples, which (I assume), are more common use case than Array subclasses in promise methods.

[Example] (https://www.typescriptlang.org/play?target=7#code/CYUwxgNghgTiAEkoGdnwIIxlAngYQAsBLCYAHgBUA+eEADwBcQA7YNTbHSmgbwF8AUANBI4iAPbNkDRMVIAuDFlyES5AAoxxAWyLIQZZgFdtAIxAwqVISOhiAZkeZgGRSfCgQIlWoxZt4TR09AycAa2ZxAHdmKgBtAF0qAAoAN08jEGRFCgBKRSDdfR96JlY0QpCyImZ7C3gAZXiE+AB+RsT4RWYQVItrW1gER2dXd08IBpAGBggQcgpfMoDK4vDImOaU9IhM7Pg8gq0ig0XS-wrjqpq6mEbmtsCr-SmZueAAJSyjCAYyJs63V6-SEE2ScVWIAAdHBkOIIH1kgBGXIAGiewX0MKy8MRAHIAI5REB43IJXICeBU+BQhgEFjJMDwAC8NDAFIA9BzqTzefAAHqtIQCSFQsEQ57Q2G4kDItEYk7YuEI2WE4mk8mU6m0+nMRkstmc7l8vmCoA)

**Comment by andrewbranch** [comment]
Yeah, we definitely can’t break tuple inference here. Seems like there could be another way to write it, but we’ll have to weigh the complexity and performance cost against how useful it actually is on average in real code.

**Comment by asnaeb** [comment]
Extending Array class, while uncommon is part of the language and this bug can be a wall against libraries development, for example. I'm going to spend some time in finding a solution as soon as I get the chance!

**Comment by asnaeb** [comment]
While I didn't test this exhaustively, It looks like it's working
```ts
declare function all<T extends unknown[]>(values: T): Promise<
    T extends {[P in 0]: unknown} ? 
        {-readonly [P in keyof T]: Awaited<T[P]>} : 
            T extends ArrayLike<infer S> | Iterable<infer S> ? Awaited<S>[] : never 
>
```
[Playground Link](https://www.typescriptlang.org/play?target=7#code/CYUwxgNghgTiAEkoGdnwIIxlAngYQAsBLCYAHgBUA+eEADwBcQA7YNTbHSmgbwF8AUANBI48AGYBXZmAZEA9s3hQIESrUYs28aQGtm8gO7MA2gF0qACgBuKySGQAueBQCUzgAox5AWyLIQMgF4EJcNJlY0HhMPeCIlAAYzZz0DYz54AH54YNC8ngBaOChgRQgceBi4pV0QHHlxF2SMQygiJnIKGIsM51y8gbD6CO0OXAAZIlqyePEQGHgAZRoAH3gASSZsACMIQNn5pZps9Fb2kHJl83hnZhBrQ4EqIRFoMTBFZAZlLFxPbz8ATIzEkPm28yo5mE4DeCA+zC+8AYkgADntnDEAf5AiCwRCADTwLy+bFkL4weIAcyohOJgMC23k8j2UGYVDMOVesDhn2+YGIpGcY3wAvIdNJ5KpVGeAhUEEssE4rn68AAdAwCCxLGB4ABeGhgZUAeiNgzNAD1MkI5ZZkWiQMq8urNcxtXqDcbTWaBpbraptaLHaFnVqdfrEJ7vT6rUJxQFVTbFbggyEQ66wx6BCao3lfQI4yAE-67XsU2qNaH3RGs16c-A8wWi-L+SRgGW027w4aa3WQpagA)
